wgpu-hal/src/dx12/device.rs

```diff
@@ -221,62 +222,151 @@ impl super::Device {
+        let load_shader_fxc = || {
+            let mut shader_data = native::Blob::null();
+            let mut compile_flags = d3dcompiler::D3DCOMPILE_ENABLE_STRICTNESS;
+
+            if self
+                .private_caps
+                .instance_flags
+                .contains(crate::InstanceFlags::DEBUG)
+            {
+                compile_flags |=
+                    d3dcompiler::D3DCOMPILE_DEBUG | d3dcompiler::D3DCOMPILE_SKIP_OPTIMIZATION;
+            }
+
+            let full_stage = format!(
+                "{}_{}\0",
+                naga_stage.to_hlsl_str(),
+                hlsl::ShaderModel::V5_1.to_str()
+            );
+            let mut error = native::Blob::null();
+
+            let hr = unsafe {
+                profiling::scope!("d3dcompiler::D3DCompile");
+                d3dcompiler::D3DCompile(
+                    source.as_ptr().cast(),
+                    source.len(),
+                    source_name.as_ptr().cast(),
+                    ptr::null(),
+                    ptr::null_mut(),
+                    raw_ep.as_ptr(),
+                    full_stage.as_ptr().cast(),
+                    compile_flags,
+                    0,
+                    shader_data.mut_void().cast(),
+                    error.mut_void().cast(),
+                )
+            };
+
+            match hr.into_result() {
+                Ok(()) => (Ok(super::ShaderDXIL::Fxc(shader_data)), log::Level::Info),
+                Err(e) => {
+                    let mut full_msg = format!("D3DCompile error ({})", e);
+                    if !error.is_null() {
+                        use std::fmt::Write as _;
+                        let message = unsafe {
+                            slice::from_raw_parts(
+                                error.GetBufferPointer() as *const u8,
+                                error.GetBufferSize(),
+                            )
+                        };
+                        let _ = write!(full_msg, ": {}", String::from_utf8_lossy(message));
+                        unsafe {
+                            error.destroy();
+                        }
+                    }
+                    (
+                        Err(crate::PipelineError::Linkage(stage_bit, full_msg)),
+                        log::Level::Warn,
+                    )
+                }
+            }
         };
 
-        let hr = unsafe {
-            profiling::scope!("d3dcompiler::D3DCompile");
-            d3dcompiler::D3DCompile(
-                source.as_ptr() as *const _,
-                source.len(),
+        // hassle_rs doesn't check for `dxil.dll` if you only call `hassle_rs::compile_hlsl` but calling hassle_rs::Dxil::new will .
+        profiling::scope!("hassle_rs::compile_hlsl");
+        let (result, log_level) = match hassle_rs::Dxil::new(None).and_then(|_| {
+            hassle_rs::compile_hlsl(
```

Comment:
We shouldn't be using `compile_hlsl` directly as it needs to load the library every single time you call the function. Ideally you'll load a DXCCompiler and DXCLibrary once on dx12 instance creation, then re-use it for every time we need to compile stuff.